handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1943,14 +1955,24 @@ public void run() {
      * @return {@code true} if {@link #handshakePromise} was set successfully and a {@link SslHandshakeCompletionEvent}
      * was fired. {@code false} otherwise.
      */
-    private boolean setHandshakeSuccess() {
+    private boolean setHandshakeSuccess() throws SSLException {
         // Our control flow may invoke this method multiple times for a single FINISHED event. For example
         // wrapNonAppData may drain pendingUnencryptedWrites in wrap which transitions to handshake from FINISHED to
         // NOT_HANDSHAKING which invokes setHandshakeSuccess, and then wrapNonAppData also directly invokes this method.
+        SSLSession session = engine.getSession();
+        if (resumptionController != null) {
+            try {
+                if (resumptionController.validateResumeIfNeeded(engine) && logger.isDebugEnabled()) {
+                    logger.debug("{} Resumed and reauthenticated session (handshake_done={})", ctx.channel(),
+                            handshakePromise.isDone());
+                }
+            } catch (CertificateException e) {
+                throw new SSLException(e);
+            }
+        }
         final boolean notified;
         if (notified = !handshakePromise.isDone() && handshakePromise.trySuccess(ctx.channel())) {
```

Comment:
@normanmaurer One thing I'm wondering if whether the resumption block should be behind this handshake-is-done gate. The code implies that the handshake promise could already be completed at this point, but I can't think of a case where that would be true, or what it would mean.

Reply:
yeah sounds more robust to me... that said this would also mean that the promise would be notified before we validate which is a bit "odd" if people expect that all the data etc is restored.

Reply:
I definitely want the resumption call to happen before the promise listeners are notified. Some frameworks look for the pipeline event, while others listen on the handshake promise. They need to behave the same in this regard...

Reply:
I've looked at two approaches here:

One where we add a `trySuccess` method that also takes a `Runnable` callback which will be run before the listeners, if we successfully change the state of the promise. To make this atomic, the runnable callback will have to be executed while holding the lock on the promise, so no listeners can be added or notified.

Another where we add a `trySuccess` method which returns a "lease" object iff we successfully set the result of the promise to this lease object. When a promise has a lease result, is not "done" and cannot be completed by anyone else. The lease object is runnable, and will set the result of the promise to the real result and notify the listeners, when run. This is more complicated, but we won't need to hold a lock on the promise for an extended period.

Reply:
Oh, I realized that we _have_ to try calling the `resume` methods before completing the handshake promise, as otherwise we can't fail it if the trust manager throws an exception.